modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/near/GridNearTransactionalCache.java

```diff
@@ -627,6 +652,166 @@ else if (!F.isEmpty(req.keys()))
         }
     }
 
+    /** {@inheritDoc} */
+    @Override public void unlockAllForSavepoint(@Nullable Collection<IgniteTxKey> keys, GridCacheVersion cacheVer,
+        GridNearTxLocal tx) throws IgniteCheckedException {
+        if (F.isEmpty(keys))
+            return;
+
+        GridCacheVersion ver = null;
+
+        int keyCnt = -1;
+
+        Map<ClusterNode, GridNearUnlockRequest> map = null;
+
+        Collection<KeyCacheObject> locKeys = new LinkedList<>();
+
+        Map<GridNearUnlockRequest, SavepointUnlockFuture> unlockFuts = new HashMap<>();
+
+        for (IgniteTxKey key : keys) {
+            while (true) {
+                KeyCacheObject cacheKey = key.key();
+
+                GridDistributedCacheEntry entry = peekExx(cacheKey);
+
+                if (entry == null)
+                    break; // While.
+
+                try {
+                    GridCacheMvccCandidate cand = entry.candidate(ctx.nodeId(), Thread.currentThread().getId());
```

Comment:
org.apache.ignite.internal.processors.cache.GridCacheMapEntry#candidate(java.util.UUID, long) method is excessive(and likely will be deleted),
you should use org.apache.ignite.internal.processors.cache.GridCacheMapEntry#localCandidate instead